storage/innobase/buf/buf0buf.cc

```diff
@@ -2959,6 +2824,47 @@ buf_block_t *buf_pool_t::page_fix(const page_id_t id,
   }
 }
 
+TRANSACTIONAL_TARGET
+uint32_t buf_pool_t::page_guess(buf_block_t *b, page_hash_latch &latch,
+                                const page_id_t id) noexcept
+{
+  transactional_shared_lock_guard<page_hash_latch> g{latch};
+#ifdef __linux__
+  /* On Linux, madvise(MADV_DONTNEED) on an anonymous mapping is
+  documented to retain the addresses as valid but pointing to
+  zero-initialized memory.  This may lead to a false positive for
+  b->page.id() == id but never for b->page.state().  At the time
+  shrunk() was invoked, shrink() and buf_LRU_block_free_non_file_page()
+  should guarantee that b->page.state() == buf_page_t::NOT_USED
+  for all to-be-freed blocks. */
+#else
+  /* shrunk() made the memory inaccessible. */
+  if (UNIV_UNLIKELY(reinterpret_cast<char*>(b) >= memory + size_in_bytes))
+    return 0;
+#endif
+  const page_id_t block_id{b->page.id()};
+#ifdef __linux__
+  /* shrunk() may have invoked MEM_UNDEFINED() on this memory. */
+  MEM_MAKE_DEFINED(&block_id, sizeof block_id);
+#endif
+
+  if (id == block_id)
+  {
+    uint32_t state= b->page.state();
+#ifdef __linux__
+    /* shrunk() may have invoked MEM_UNDEFINED() on this memory */
+    MEM_MAKE_DEFINED(&state, sizeof state);
```

Comment:
Again, `#if __linux__` . please avoid. The whole functionality is not `__linux__` only, and needs to be well tested elsewhere.